### ISSUE TITLE
Research and convert uses of Thread.Sleep() to Task.Wait() where applicable

### ISSUE THREAD

**Issue by cabal95** [SparkDevNetwork/Rock#3125]
# Feature Request

**Describe the Feature Request**

In cases where something is running as a background task and needs to pause execution, the `Task.Wait()` method should be used instead of the current `Thread.Sleep()`. This is primarily because a single thread can actually run more than one task concurrently.

If a `Task` is waiting on a network operation, the entire thread is not paused. Instead the worker thread will process another `Task` while waiting on the first task's network operation. When `Thread.Sleep()` is called in this case, it forcibly pauses the entire thread, which could pause other tasks.

@nairdo FYI.

**Describe Preferred Solution**

In cases where something is running as a background task, `Task.Wait()` should be used instead of `Thread.Sleep()`.

**Describe Alternatives**

We could build an entirely new task library from the ground up that intercepts calls to `Thread.Sleep` and replaces them with `Task.Wait` by using Reflection and code generation tools - but that seems like a harder and more wasteful solution than just switching to Task.Wait. :)

**Providing Code**
I will provide code if feature request is approved: **Yes**

**Comment by cabal95** [comment]
"Just switch" meaning, don't worry about the "research" part of the issue title - correct?

**Comment by IGx89** [comment]
I'm late, just randomly came across this and am not a Rock developer, but Task.Wait() will actually block the thread too (it's a blocking operation). What you describe, letting a thread process other tasks while waiting for the delay to pass, requires using the `await` keyword instead. That's a much bigger change (for instance, it looks like you'd have to update your old, custom build of Quartz to their latest v3 to use async/await in your Jobs code), but still definitely worth it in the long run.
